platforms/software/ivy/src/main/java/org/gradle/api/publish/ivy/internal/artifact/AbstractIvyArtifact.java

```diff
@@ -16,95 +16,86 @@
 
 package org.gradle.api.publish.ivy.internal.artifact;
 
-import com.google.common.base.Strings;
 import org.gradle.api.internal.tasks.DefaultTaskDependency;
 import org.gradle.api.internal.tasks.TaskDependencyFactory;
+import org.gradle.api.model.ObjectFactory;
+import org.gradle.api.provider.Property;
+import org.gradle.api.provider.Provider;
+import org.gradle.api.provider.ProviderFactory;
+import org.gradle.api.tasks.Optional;
 import org.gradle.api.tasks.TaskDependency;
 
-import javax.annotation.Nullable;
-
 public abstract class AbstractIvyArtifact implements IvyArtifactInternal {
     private final TaskDependency allBuildDependencies;
     private final DefaultTaskDependency additionalBuildDependencies;
-
-    private String name;
-    private String type;
-    private String extension;
-    private String classifier;
-    private String conf;
-
-    protected AbstractIvyArtifact(TaskDependencyFactory taskDependencyFactory) {
+    private final Property<String> nameProperty;
+    private final Property<String> typeProperty;
+    private final Property<String> extensionProperty;
+    private final Property<String> classifierProperty;
+    private final Property<String> confProperty;
+
+    protected AbstractIvyArtifact(
+        TaskDependencyFactory taskDependencyFactory,
+        ProviderFactory providerFactory,
+        ObjectFactory objectFactory
+    ) {
         this.additionalBuildDependencies = new DefaultTaskDependency();
         this.allBuildDependencies = taskDependencyFactory.visitingDependencies(context -> {
             context.add(getDefaultBuildDependencies());
             additionalBuildDependencies.visitDependencies(context);
         });
-    }
-
-    @Override
-    public String getName() {
-        return name != null ? name : getDefaultName();
-    }
-
-    protected abstract String getDefaultName();
 
-    @Override
-    public void setName(String name) {
-        this.name = Strings.nullToEmpty(name);
+        this.nameProperty = objectFactory.property(String.class);
+        this.typeProperty = objectFactory.property(String.class);
+        this.extensionProperty = objectFactory.property(String.class);
+        this.classifierProperty = objectFactory.property(String.class);
+        this.confProperty = objectFactory.property(String.class);
```

Comment:
Any reason we're not using abstract getters for these properties?

Reply:
Then should DerivedIvyArtifact even extend AbstractIvyArtifact? Or even if it did, couldn’t it just implement the abstract getters anyways

Reply:
I think redesigning the class hierarchy is out of scope of this PR

Reply:
I agree we shouldn't be doing any redesigning here. 

Looking at DerivedIvyArtifact, it doesn't look like its implementing any of these properties